source/rust_verify/example/guide/modes.rs

```diff
@@ -381,6 +381,73 @@ fn add_one(x: u8) -> (ret: u8)
 }
 // ANCHOR_END: const1
 
+// ANCHOR: fib 
+spec fn fib(n: nat) -> nat
+    decreases n,
+{
+    if n == 0 {
+        0
+    } else if n == 1 {
+        1
+    } else {
+        fib((n - 2) as nat) + fib((n - 1) as nat)
+    }
+}
+
+proof fn lemma_fib_is_monotonic(i: nat, j: nat)
+    requires
+        i <= j,
+    ensures
+        fib(i) <= fib(j),
+    decreases j - i,
+{
+    if i < 2 && j < 2 {
+    } else if i == j {
+    } else if i == j - 1 {
+        assert(fib(j) == fib((j - 2) as nat) + fib((j - 1) as nat));
+        lemma_fib_is_monotonic(i, (j - 1) as nat);
+    } else {
+        lemma_fib_is_monotonic(i, (j - 1) as nat);
+        lemma_fib_is_monotonic(i, (j - 2) as nat);
+    }
+}
+
+spec fn fib_fits_u64(n: nat) -> bool {
+    fib(n) <= 0xffff_ffff_ffff_ffff
+}
+
+exec fn fib_impl(n: u64) -> (result: u64)
+    requires
+        fib_fits_u64(n as nat),
+    ensures
+        result == fib(n as nat),
```

Comment:
This version of the fibonacci demo is a bit more compact, but sort-of unrealistic, due to the fact that the caller needs to prove that the result will fit in a `u64`. I have a more realistic version that returns an `Option` that's `None` if the result doesn't fit: what do you think, shall I replace it on this branch?

Reply:
I think fibonacci is always going to be rather artificial, unfortunately.  We have more realistic [examples](https://github.com/secure-foundations/human-eval-verus/blob/4196755da573deb08c3cff684dbfe4163ddbe961/tasks/human_eval_055b.rs) in HumanEval, but at the end of the day, it's still a bit silly, so I think there may be some virtue in keeping the example simple to focus on what it takes to combine the different modes together.